### DISCUSSION TITLE
Indexing vertex or edge labels

### DISCUSSION THREAD

**Discussion by marco-brandizi** [ArcadeData/arcadedb#1136] [Q&A] [answered]
Hi all again, 

According to the syntax of `CREATE INDEX`, it appears that creating an index on vertex or edge labels isn't possible in ArcadeDB. Would you confirm that? Are label searches still efficient? That is, are queries like these fast enough:

```
# Gremlin
g.V().hasLabel ( 'Person' )
g.V().hasLabel ( containing ( 'Person' )  )

# Cypher
match (n) where labels (n)[0] = 'Person' return n
match (n) where labels (n)[0] contains 'Person' return n
```

From my experiments, Gremlin versions seem quite fast (10ms on my laptop, with a few thousand element database), Cypher queries are quite slow (1s). By the way, `EXPLAIN` doesn't work with Cypher and when I use the `explain()` step with Gremlin, I'm told Gremlin-level query plan optimisations, but nothing about the database engine.

**Comment by lvca** [comment, marked as answer]
Hi, we don't use indexes for labels, but rather use physical separated files to store them. This is much more efficient than using indexes. The issue with the contains is that the optimizer doesn't resolve as a fetch from type like the one above.

**Reply by marco-brandizi** [reply]
Thanks, good to know!

**Reply by lvca** [reply]
By the way, with Cypher if you want to go faster, use this syntax instead: `match (n:Person) return n`. It should be comparable to the Gremlin performance.

**Reply by marco-brandizi** [reply]
> use this syntax instead: match (n:Person) return n
>
I was testing the case where the label is like: `Person:Employee:Manager` (the way to emulate multiple labels where they aren't supported), and I want to search for nodes of type Person, without knowing the other labels they might have.

In Gremlin, the `containing()` operator seems to work as fast as the exact match, in Cypher, using labels() looks quite slow.
